### ISSUE TITLE
[Discussion] Filters 

### ISSUE THREAD

**Issue by Mimetis** [Mimetis/Dotmim.Sync#179]
Let's discuss how to improve filters implementation

**Comment by lordofffarm** [comment]
Did you intend for `FilterClause2` to be something like this?
![Screenshot_2020-01-09_18-12-04](https://user-images.githubusercontent.com/28428684/72066813-b9052380-330b-11ea-942d-0430e44f5a2d.png)

**Comment by Mimetis** [comment]
No I think the tests I made with the PR #99 are too complicated.
I'm thinking about something different right now, but nothing definitive so far.

@gentledepp has worked on a solution you can see here : https://github.com/gentledepp/SyncFilters

I still searching for a solution that can fit to any situation, but it's quite complicated :)

**Comment by Mimetis** [comment]
Here is a first draft on what I m thinking about.

## Assumptions

We have 3 tables, on which we want to filter by `customerId`:

- `Customer` : `CustomerId` (guid) - `Name` (string)
- `CustomerAddress` : `CustomerAddressId` (guid) - `AddressType` (enum string) - `AddressId` (guid) - `CustomerId` (guid)
- `Address` : `AddressId` (guid) - `AddressLine1` (string) - `City` (string)

![image](https://user-images.githubusercontent.com/4592555/73168848-07336880-40fb-11ea-900f-9e574d19b01e.png)

As you can see, we can have multiple addresses for one customer.

**TODAY**, we can create two `SetupFilter` on table `Customer` and `CustomerAddress`, using this kind of code:
``` cs
setup.Filters.Add("Customer", "CustomerID");
setup.Filters.Add("CustomerAddress", "CustomerID");
```
The SQL code generated for the `_Changes` stored procedure looks like:
``` sql
SELECT 	[side].[CustomerID], 
	[base].[EmployeeID],  [base].[NameStyle],  [base].[FirstName],  [base].[LastName], 
	[side].[sync_row_is_tombstone],  [side].[update_scope_id] 
FROM [Customer] [base]
RIGHT JOIN [Customer_tracking] [side] ON [base].[CustomerID] = [side].[CustomerID]
WHERE (
	([side].[CustomerID] = @CustomerID) OR ([side].[CustomerID] IS NULL)
	)
AND (
	[side].[timestamp] > @sync_min_timestamp AND ([side].[update_scope_id] <> @sync_scope_id OR [side].[update_scope_id] IS NULL)
)
```
**BUT** we **can't** create a `SetupFilter` on table `Address` because there is no reference / column `CustomerId` referencing the table `Customer`...

## Solution

I'm thinking about this kind of new filter, where we will be allowed to add some sort of *custom joins and where* text.
Regarding the `Address_changes` stored procedure, we can imagine something like this: 

This kind of code:
``` cs
addressFilter.AddLeftJoin("CustomerAddress").On("CustomerAddress.AddressId", "side.AddressId")
 .AddLeftJoin("Customer").On("Customer.CustomerID", "CustomerAddress.CustomerID")
 .AddWhereFilterEquals("Customer.CustomerId")
```
Will generate this SQL code:
``` sql
SELECT 	[side].[AddressID], 
	[base].[AddressLine1], [base].[City], [base].[PostalCode], [side].[sync_row_is_tombstone], [side].[update_scope_id] 
FROM [Address] [base]
RIGHT JOIN [Address_tracking] [side] ON [base].[AddressID] = [side].[AddressID]
LEFT JOIN [CustomerAddress] on [CustomerAddress].[AddressId]= [side].[AddressID]
LEFT JOIN [Customer] on [Customer].[CustomerID]= [CustomerAddress].[CustomerID]

WHERE 
	(
	[Customer].[CustomerID] = @CustomerID
	)
AND (
	[side].[timestamp] > @sync_min_timestamp
	AND ([side].[update_scope_id] <> @sync_scope_id OR [side].[update_scope_id] IS NULL)
)
```
We can deeper, with this kind of code:
``` cs
addressFilter.AddLeftJoin("CustomerAddress").On("CustomerAddress.AddressId", "side.AddressId")
 .AddCustomerWhere("CustomerAddress.CustomerId in (3,4,5)");

``` 
Will generate this SQL:
``` sql
SELECT 	[side].[AddressID], 
	[base].[AddressLine1], [base].[City], [base].[PostalCode], [side].[sync_row_is_tombstone], [side].[update_scope_id] 
FROM [Address] [base]
RIGHT JOIN [Address_tracking] [side] ON [base].[AddressID] = [side].[AddressID]
LEFT JOIN [CustomerAddress] on [CustomerAddress].[AddressId]= [side].[AddressID]
WHERE 
	(
		CustomerAddress.CustomerId in (3,4,5)
	)
AND (
	[side].[timestamp] > @sync_min_timestamp
	AND ([side].[update_scope_id] <> @sync_scope_id OR [side].[update_scope_id] IS NULL)
)
```
What do you think guys ?

cc @gentledepp : I know you've a lot of experiences on filters, is this kind of solution could fit to your scenario ?

**Comment by lordofffarm** [comment]
This is what I had in mind too. Seems to be a decent solution.

**Comment by lordofffarm** [comment]
There is one issue with filtering on columns though.

That is if the values in the filtered columns change.

For example: let's say the filter is on the column FirstName of the Customer table.
The filtering value is set to "John".
When synchronizing, clients will get rows with the FirstName = John.
Then one "customer" decides to change his FirstName to "Jonathan"
When synchronizing , clients will not get the update to the row since the filtering value is set to "John".
Which will result in a stale row on client. 

I feel like this is resolved in @gentledepp 's solution in https://github.com/gentledepp/SyncFilters

SymmetricDS has something similar (I feel like) 
https://www.symmetricds.org/doc/3.9/html/user-guide.html#_channels
which probably resolves this particular issue as well.

What do you think about the "Channels" solution?

**Comment by Mimetis** [comment]
@lordofffarm : Indeed, if the filter value changes, it won't work.
I will try to go a little bit deeper into the symmetricds channels system, because I've never used this framework so far.

For now, I've implemented a new `SyncFilter` solution, you can test in the last commit.
## How It works

Now you can apply a filter on any table, even if the column belongs to another table.
For instance, you can apply a filter on the **Customer** table, even if the filter is on the **Address** table (for example, filtering on the **City**)

*In a nutshell*, the new system add a new layer of **flexibility** but at the cost of a new layer of **complexity** :)

To illustrate how it works, here is a straightforward scenario:

*We Want only **Customer** from specific **City** and a specific **Postal code**.*
*Each customer has **Addresses** and **Sales Orders** which should be filtered as well.*

![image](https://user-images.githubusercontent.com/4592555/73394907-c98f3500-42de-11ea-85ae-3b242585b88e.png)

We will have to filter:
- Level zero: **Address**
- Level one: **CustomerAddress**
- Level two: **Customer**, **SalesOrderHeader**
- Level four: **SalesOrderDetail**

So far, here is the new methods on the `SetupFilter` you have to achieve this:

### `SetupFilter` class
The `SetupFilter` class will allow you to personalize your filter on a defined table (`Customer` in this example)
``` cs
var customerFilter = new SetupFilter("Customer");
```
Be careful, you can have only one `SetupFilter` instance per table. 
Obviously, this instance will allow you to define multiple parameters / criterias 

### `.AddParameter()`
Allows you to add a new parameter to the `_changes` stored procedure.
This method can be called with two kind of arguments:
- Your parameter is a **custom** parameter. You have to define its name and its `DbType`. Optionally, you can define if it can be null and its default value (SQL Server only)
- Your parameter is a **mapped**  column. Easier, you just have to define its name and the mapped column. This way, `Dotmim.Sync` will determine the parameter properties, based on the schema

For instance, the parameters declaration for the table `Customer` looks like:
``` cs
customerFilter.AddParameter("City", "Address", true);
customerFilter.AddParameter("postal", DbType.String, true, null, 20);

```
- `City` parameter is defined from the `Address`.`City` column
- `postal` parameter is a custom defined parameter 
   - *Indeed we have a `PostalCode` column in the `Address` table, that could be used here. But we will use a custom parameter instead, for the example*

At the end, the generation code should looks like:

``` sql
ALTER PROCEDURE [dbo].[sCustomerAddress_Citypostal__changes]
	@sync_min_timestamp bigint,
	@sync_scope_id uniqueidentifier,
	@City varchar(MAX) NULL,
	@postal nvarchar(20) NULL
```
Where `@City` is a mapped parameter and `@postal` is a custom parameter.

### `.AddJoin()`

If your filter is applied on a column in the actual table, you don't need to add any `join` statement.
But, in our example, the `Customer` table is two levels below the `Address` table (where we have the filtered columns `City` and `PostalCode`)
So far, we can add some join statement here, going from `Customer` to `CustomerAddress` then to `Address`:
``` cs
customerFilter.AddJoin(Join.Left, "CustomerAddress").On("CustomerAddress", "CustomerId", "Customer", "CustomerId");
customerFilter.AddJoin(Join.Left, "Address").On("CustomerAddress", "AddressId", "Address", "AddressId");
```
The generated statement now looks like:
``` sql
FROM [Customer] [base]
RIGHT JOIN [tCustomer] [side]ON [base].[CustomerID] = [side].[CustomerID]
LEFT JOIN [CustomerAddress] ON [CustomerAddress].[CustomerId] = [base].[CustomerId]
LEFT JOIN [Address] ON [CustomerAddress].[AddressId] = [Address].[AddressId]
```
As you can see the `Dotmim.Sync` framework will take care of quoted table / column names and aliases in the stored procedure.
Just focus on the name of your table.


### `.AddWhere()`

Now, and for each parameter, you will have to define the where condition
Each parameter will be compare to an existing column in an existing table.
For instance: 
- The `City` parameter should be compared to the `City` column in the `Address` table.
- The `postal` parameter should be compared to the `PostalCode` column in the `Address` table:

```cs
// Mapping City parameter to Address.City column
addressFilter.AddWhere("City", "Address", "City");
// Mapping the custom "postal" parameter to Address.PostalCode
addressFilter.AddWhere("PostalCode", "Address", "postal");

```
The generated sql statement looks like this:
``` sql
WHERE (
(
 (
   ([Address].[City] = @City OR @City IS NULL) AND ([Address].[PostalCode] = @postal OR @postal IS NULL)
  )
 OR [side].[sync_row_is_tombstone] = 1
)
```
### `.AddCustomWhere()`
If you need more, this method will allow you to add your own where condition.
Be careful, this method takes a `string` as argument, which will not be parsed, but instead, just added at the end of the stored procedure statement.

## Example

Here is the full sample, where we define the filters (`City` and `postal` code) on each filtered tables: `Customer`, `CustomerAddress`, `Address`, `SalesOrderHeader` and `SalesOrderDetail`

You will find the source code in the last commit, project `Dotmim.Sync.SampleConsole.csproj`, file `program.cs`, method `SynchronizeAsync()`:

``` cs
var setup = new SyncSetup(new string[] {"ProductCategory",
  "ProductModel", "Product",
  "Address", "Customer", "CustomerAddress",
  "SalesOrderHeader", "SalesOrderDetail" });

// ----------------------------------------------------
// Horizontal Filter: On rows. Removing rows from source
// ----------------------------------------------------
// Over all filter : "we Want only customer from specific city and specific postal code"
// First level table : Address
// Second level tables : CustomerAddress
// Third level tables : Customer, SalesOrderHeader
// Fourth level tables : SalesOrderDetail

// Create a filter on table Address on City Washington
// Optional : Sub filter on PostalCode, for testing purpose
var addressFilter = new SetupFilter("Address");

// For each filter, you have to provider all the input parameters
// A parameter could be a parameter mapped to an existing colum : That way you don't have to specify any type, length and so on ...
// We can specify if a null value can be passed as parameter value : That way ALL addresses will be fetched
// A default value can be passed as well, but works only on SQL Server (MySql is a damn shity thing)
addressFilter.AddParameter("City", "Address", true);

// Or a parameter could be a random parameter bound to anything. In that case, you have to specify everything
// (This parameter COULD BE bound to a column, like City, but for the example, we go for a custom parameter)
addressFilter.AddParameter("postal", DbType.String, true, null, 20);

// Then you map each parameter on wich table / column the "where" clause should be applied
addressFilter.AddWhere("City", "Address", "City");
addressFilter.AddWhere("PostalCode", "Address", "postal");
setup.Filters.Add(addressFilter);

var addressCustomerFilter = new SetupFilter("CustomerAddress");
addressCustomerFilter.AddParameter("City", "Address", true);
addressCustomerFilter.AddParameter("postal", DbType.String, true, null, 20);

// You can join table to go from your table up (or down) to your filter table
addressCustomerFilter.AddJoin(Join.Left, "Address").On("CustomerAddress", "AddressId", "Address", "AddressId");

// And then add your where clauses
addressCustomerFilter.AddWhere("City", "Address", "City");
addressCustomerFilter.AddWhere("PostalCode", "Address", "postal");
setup.Filters.Add(addressCustomerFilter);

var customerFilter = new SetupFilter("Customer");
customerFilter.AddParameter("City", "Address", true);
customerFilter.AddParameter("postal", DbType.String, true, null, 20);
customerFilter.AddJoin(Join.Left, "CustomerAddress").On("CustomerAddress", "CustomerId", "Customer", "CustomerId");
customerFilter.AddJoin(Join.Left, "Address").On("CustomerAddress", "AddressId", "Address", "AddressId");
customerFilter.AddWhere("City", "Address", "City");
customerFilter.AddWhere("PostalCode", "Address", "postal");
setup.Filters.Add(customerFilter);

var orderHeaderFilter = new SetupFilter("SalesOrderHeader");
orderHeaderFilter.AddParameter("City", "Address", true);
orderHeaderFilter.AddParameter("postal", DbType.String, true, null, 20);
orderHeaderFilter.AddJoin(Join.Left, "CustomerAddress").On("CustomerAddress", "CustomerId", "SalesOrderHeader", "CustomerId");
orderHeaderFilter.AddJoin(Join.Left, "Address").On("CustomerAddress", "AddressId", "Address", "AddressId");
orderHeaderFilter.AddWhere("City", "Address", "City");
orderHeaderFilter.AddWhere("PostalCode", "Address", "postal");
setup.Filters.Add(orderHeaderFilter);

var orderDetailsFilter = new SetupFilter("SalesOrderDetail");
orderDetailsFilter.AddParameter("City", "Address", true);
orderDetailsFilter.AddParameter("postal", DbType.String, true, null, 20);
orderDetailsFilter.AddJoin(Join.Left, "SalesOrderHeader").On("SalesOrderHeader", "SalesOrderID", "SalesOrderHeader", "SalesOrderID");
orderDetailsFilter.AddJoin(Join.Left, "CustomerAddress").On("CustomerAddress", "CustomerId", "SalesOrderHeader", "CustomerId");
orderDetailsFilter.AddJoin(Join.Left, "Address").On("CustomerAddress", "AddressId", "Address", "AddressId");
orderDetailsFilter.AddWhere("City", "Address", "City");
orderDetailsFilter.AddWhere("PostalCode", "Address", "postal");
setup.Filters.Add(orderDetailsFilter);

// ----------------------------------------------------
```
And you `SyncAgent` now looks like:

``` cs
// Creating an agent that will handle all the process
var agent = new SyncAgent(clientProvider, serverProvider, setup);

if (!agent.Parameters.Contains("City"))
    agent.Parameters.Add("City", "Toronto");

// Because I've specified that "postal" could be null, I can set the value to DBNull.Value (and the get all postal code in Toronto city)
if (!agent.Parameters.Contains("postal"))
    agent.Parameters.Add("postal", DBNull.Value);

var s1 = await agent.SynchronizeAsync(progress);

```

**Comment by mdonogma** [comment]
Oh my , you guys are amazing! I’ve been performance testing 0.4 to and it’s a massive improvement over 0.3

**Comment by Mimetis** [comment]
Hey @mdonogma !
Thanks for the feedback !
I'm really interested about your protocol performance test ?
How did you realize a test comparision between `v0.3` and `v0.4` ?

**Comment by mdonogma** [comment]
Ran 10 client db’s and 1 server dB.

Clients would insert 1000 rows then sync with servers

Clients were run simultaneously to ensure concurrency.

Just kept going till about 1million records in table.

Tested and 0,3 then 0,4 - 0.4 won hands down

**Comment by mdonogma** [comment]
I’ll be more scientific about it when 0.4 ships. These tests were against webapi as sync server will be hosted on aws.

**Comment by Mimetis** [comment]
`v0.4` was shipped a few hours ago :)

Which providers did you use during your tests ?
`SqlSyncChangeTrackingProvider` has been released with the `v0.4`. Did you tried it ?

**Comment by mdonogma** [comment]
I’ll give the sql change provider then run tests again.

Sent from my iPhone

On 2 Feb 2020, at 19:50, Sébastien Pertus <notifications@github.com> wrote:

﻿

v0.4 was shipped a few hours ago :)

Which providers did you use during your tests ?
SqlSyncChangeTrackingProvider has been released with the v0.4. Did you tried it ?

**Comment by mdonogma** [comment]
Fantastic work on 0.4 - super excited !

Sent from my iPhone

On 2 Feb 2020, at 19:50, Sébastien Pertus <notifications@github.com> wrote:

﻿

v0.4 was shipped a few hours ago :)

Which providers did you use during your tests ?
SqlSyncChangeTrackingProvider has been released with the v0.4. Did you tried it ?

**Comment by Mimetis** [comment]
Let me know if everything is working correctly with the change tracking provider.
I don't have any feedback with this provider, since it's a new one.
So need your thoughts / results :)

**Comment by gentledepp** [comment]
@mdonogma 
**Watch out: 0.4 is not backwards compatible with 0.3!!** 
As far as I am aware (I was on holidays the last week) there is no migration guide from 0.3 to 0.4 so be very, very careful!

Also @Mimetis: Your syncfilter solution does not work for me.
1. as @lordofffarm  pointed out, it fails whenever the columns that are used to filter, change
2. moreover, it fails, whenever there is any change in the relationship (in the tables in bewteen) that would affect the result set
3. moreover, you are directly joining other tables, thereby muplitplying the rows in your result set (ponentially). For example: With in `sp_selectchanges_customeraddress` you could get the same cusomeraddress **multiple times** because of the joins!

So I'm afraid your approach will lead to a dead-end. :-|

**Comment by Mimetis** [comment]
Hey @gentledepp 

Nice to see you here again :)

### Fails whenever the columns that are used to filter, change

Yes, right ! My advice is to sync on primary columns instead of any other potential "changing column value"

### Fails, whenever there is any change in the relationship (in the tables in bewteen) that would affect the result set

Anyway, If **anything** in the database schema is changing, the whole sync has to be redefined (using `DeprovisionAsync()` and `ProvisionAsync()`

Using `DeprovisionAsync()` and `ProvisionAsync()`, you can update a unique table if needed:
Let's suppose you're modifying a column in the table `Customer`. You can use something like:
``` csharp
SqlSyncProvider serverProvider = new SqlSyncProvider(GetDatabaseConnectionString("Northwind"));

// tables to provision
var tables = new string[] { "Customer" };

// Deprovision the table
await clientProvider.DeprovisionAsync(tables,  SyncProvision.StoredProcedures  | SyncProvision.TrackingTable  | SyncProvision.Triggers);

// Make the Alter table to add / remove column
// ...

// Provision again
await clientProvider.ProvisionAsync(tables,  SyncProvision.StoredProcedures  | SyncProvision.TrackingTable  | SyncProvision.Triggers);
```

### Fails, because muplitplying the rows in your result set

I know, I had this result at the begining of the dev :) The new version of the `_Changes` stored procedure is using a `Select Distinct` statement.

The new stored procedure, for a filter on a column that belongs to another table, looks like this:
``` sql
ALTER PROCEDURE [dbo].[sCustomer_Citypostal__changes]
	@sync_min_timestamp bigint = 0,
	@sync_scope_id uniqueidentifier NULL = NULL,
	@City varchar(MAX) NULL,
	@postal nvarchar(20) NULL
AS
BEGIN
SELECT DISTINCT	[side].[CustomerID], 
	[base].[NameStyle], 
	[base].[Title], 
	[base].[FirstName], 
	[base].[MiddleName], 
	[base].[LastName], 
	[base].[Suffix], 
	[base].[CompanyName], 
	[base].[SalesPerson], 
	[base].[EmailAddress], 
	[base].[Phone], 
	[base].[PasswordHash], 
	[base].[PasswordSalt], 
	[base].[rowguid], 
	[base].[ModifiedDate], 
	[side].[sync_row_is_tombstone], 
	[side].[update_scope_id] 
FROM [Customer] [base]
RIGHT JOIN [tCustomer] [side] ON [base].[CustomerID] = [side].[CustomerID]
LEFT JOIN [CustomerAddress] ON [CustomerAddress].[CustomerId] = [base].[CustomerId]
LEFT JOIN [Address] ON [CustomerAddress].[AddressId] = [Address].[AddressId]

WHERE (
(
 (
   ([Address].[City] = @City OR @City IS NULL) AND ([Address].[PostalCode] = @postal OR @postal IS NULL)
  )
 OR [side].[sync_row_is_tombstone] = 1
)
AND 
	[side].[timestamp] > @sync_min_timestamp
	AND ([side].[update_scope_id] <> @sync_scope_id OR [side].[update_scope_id] IS NULL)
)

```

**Comment by gentledepp** [comment]
Well, I'm afraid I did not explain  **Fails, whenever there is any change in the relationship (in the tables in bewteen) that would affect the result set** correctly

What I mean by this is that potentially, the data in the joined tables could change in a way that affects the resultset of your select_changes sp.

For example: The customer could move his office to a new CustomerAddress. The Address may already be existing and you just update the row in "CustomerAccess", or you add a new row in Address and update (or remove+add) the row in the CustomerAccess table.

All those changes will not be picked up by your sync logic.
**So basically, you always have to join in the tracking tables of all the tables that you join in your query!**
This results in a very, very complex query which is 
* hard to maintain
* and terribly slow

We used this approach in an older project (back in 2014) and it was really a mess. That's why I abandoned it eventually and moved on to our new "syncfilters" approach.

**Comment by Mimetis** [comment]
Yep, right.
I tried to understand your new system, but it's really tight to your business rules and your data.
Can you explain a little bit more how it's working ?

**Comment by mdonogma** [comment]
Would you mind create a simple filter example too. Something that probably every system needs is access control to relevant records.

Example: 
AccessTasksTable
       [Id]
      ,[TaskId]
      ,[UserId]

TasksTable
      [Id]
      ,[TaskId]
      ,[Title]


All we need to do is pass UserID as Guid of logged in user.
It should then filter the TasksTable based on join TaskId in both tables.

Reason I ask is your example is a little to complex to follow for beginners.

**Comment by Mimetis** [comment]
I'm not sure, but I guess you should have something like:
``` cs
// Because you have user id in AccessTasksTable, filter is easy:
setup.Filters.Add("AccessTasksTable", "UserId");

// Second table does not have UserId, so built a filter like this:
var taskTableFilter = new SetupFilter("TasksTable");
taskTableFilter.AddParameter("UserId", "AccessTasksTable", true);
taskTableFilter.AddJoin(Join.Left, "AccessTasksTable").On("AccessTasksTable", "TaskId", "TasksTable", "TaskId");
taskTableFilter.AddWhere("UserId", "AccessTasksTable", "UserId");
setup.Filters.Add(taskTableFilter);
```
Not sure I made the correct `AddJoin`, because I don't know the dif between `Id` and `TaskId` in your tables, but you get the idea

**Comment by mdonogma** [comment]
Essentially one or more users can be assigned access to a task via the task access table.

When the sync runs each client sends the logged in userid as a parameter. Then the magic happens , your sync engine ensures that the task only shows on clients who’s userid has access to the task.

This same rule can be applied to n-number of objects in the system.

We’re blown away by the performance of your sync library compared to other well know sync libraries. Keep up the great work!
Hope that makes sense.

Sent from my iPhone

On 4 Feb 2020, at 23:06, Sébastien Pertus <notifications@github.com> wrote:

﻿

Yep, right.
I tried to understand your new system, but it's really tight to your business rules and your data.
Can you explain a little bit more how it's working ?

**Comment by Mimetis** [comment]
Which other libraries did you test so far ?

**Comment by mdonogma** [comment]
So how and where do I pass the custom value to populate the userid?

Can this be control server and/or client side of needed.

Sent from my iPhone

On 5 Feb 2020, at 18:47, Sébastien Pertus <notifications@github.com> wrote:

﻿

I'm not sure, but I guess you should have something like:

// Because you have user id in AccessTasksTable, filter is easy:
setup.Filters.Add("AccessTasksTable", "UserId");

// Second table does not have UserId, so built a filter like this:
var taskTableFilter = new SetupFilter("TasksTable");
taskTableFilter.AddParameter("UserId", "AccessTasksTable", true);
taskTableFilter.AddJoin(Join.Left, "AccessTasksTable").On("AccessTasksTable", "TaskId", "TasksTable", "TaskId");
taskTableFilter.AddWhere("UserId", "AccessTasksTable", "UserId");
setup.Filters.Add(taskTableFilter);

Not sure I made the correct AddJoin, because I don't know the dif between Id and TaskId in your tables, but you get the idea

**Comment by Mimetis** [comment]
Something like that:
```cs
var agent = new SyncAgent(clientProvider, serverProvider, setup);

if (!agent.Parameters.Contains("UserID"))
    agent.Parameters.Add("UserID", 1001);

var s1 = await agent.SynchronizeAsync(progress);
```

**Comment by mdonogma** [comment]
Another filter question: Is there a way to enforce a server side filter that would apply to all users?
Example: A way to ensure Language settings sync based on the users server side preferred Language settings

**Comment by Mimetis** [comment]
Yes.
Use `CustomWhere`
It should do the trick

**Comment by mdonogma** [comment]
tested others but the two most well known didn't come close to yours in performance.
AWS AppSync - Amazon Web Services
Offline sync feature of Azure Mobile Apps

**Comment by gentledepp** [comment]
> Yep, right.
> I tried to understand your new system, but it's really tight to your business rules and your data.
> Can you explain a little bit more how it's working ?

Well, the system itself is not tailored to a specific business domain.

In essence, when you have a table `Customer` which should be synced, then Dotmim.Sync will add the tracking table `Customer_tracking` to keep track of any changes on the `Customer`table.
This gives you the changes you could send to the client.

What we have introduced is another table `Customer_syncfilter` which is used to filter the rows of the `Customer` table.
For every user that should receive a customer record C1, C2, C3, you have to add a separate syncfilter row.
E.g.: Andy and Garry should receive C1,C2,C3, then the `Customer_syncfilter` table will contain
C1_Andy
C2_Andy
C3_Andy
C1_Garry
C2_Garry
C3_Garry

The big advantage here is that we can fine-grained specify which data to send to which user. Because the changes in the `Customer_syncfilter` table are also tracked.

So when the data is to be selected for Gary, the sp_selectchanges_customer will
* Get all Customers who have changed since sync_min_timestamp
* or whose syncfilter (e.g. C2_Garry) has changed since sync_min_timestamp

Now obviously, with a lot of users this does not scale well, because there would be too many syncfilter rows in the database.
To overcome this issue, we introduced a graph-table to keep track of groups that users belong to. (and yes, the changes in the graph are also tracked!)

So you can assign Garry and Andy to a group "Usergroup01" and create a syncfilter for the group.
That already dramatically reduces the number of syncfilter rows.

Additionally, there may be different reasons for why a user should receive a specific record.
E.g. Garry should receive C1 because he is the creator of the C1 record
Additionally, Garry is set as "key account manager" of C1

So there are basically 2 reasons why Garry should receive C1.
When one of those reasons goes away (e.g. Andy is the new "key account manager" of C1, the system needs to know whether the syncrule can still be kept in place or if it has to be removed.
Therefore, for each sync reason - a separate syncrow would be required. So in this sample, this would normally result in 2 syncrows! 
C1_Garry_OwnerOfRecord
C1_Garry_IsKeyAccountManager


Again, we need to keep the number of syncrows as small as possible.
Therefore we introduced a bigint column "AppliedRules", which is used as a Flag-Enum store.

```csharp
[Flags]
public enum CustomerSyncReason `
{
   OwnerOfRecord = 1,
   IsKeyAccountManager = 2,
   SomeOtherReason = 4
}
```
so that way you can store up to 64 reasons per syncrow for why it should still exist.

If "AppliedRules" is zero (0), then the record C1 will be sent as toombstoned to the client.


As you see, this ruleset can be applied to almost any domain.
However, those rules have to be calculated by the application using its domain logic.
I don't see how this can be achieved by dotmim.sync!

**Comment by mdonogma** [comment]
Oh my- you guys are amazing

Sent from my iPhone

On 20 Feb 2020, at 20:32, Alexander Marek <notifications@github.com> wrote:

﻿

Yep, right.
I tried to understand your new system, but it's really tight to your business rules and your data.
Can you explain a little bit more how it's working ?

Well, the system itself is not tailored to a specific business domain.

In essence, when you have a table Customer which should be synced, then Dotmim.Sync will add the tracking table Customer_tracking to keep track of any changes on the Customertable.
This gives you the changes you could send to the client.

What we have introduced is another table Customer_syncfilter which is used to filter the rows of the Customer table.
For every user that should receive a customer record C1, C2, C3, you have to add a separate syncfilter row.
E.g.: Andy and Garry should receive C1,C2,C3, then the Customer_syncfilter table will contain
C1_Andy
C2_Andy
C3_Andy
C1_Garry
C2_Garry
C3_Garry

The big advantage here is that we can fine-grained specify which data to send to which user. Because the changes in the Customer_syncfilter table are also tracked.

So when the data is to be selected for Gary, the sp_selectchanges_customer will

  *   Get all Customers who have changed since sync_min_timestamp
  *   or whose syncfilter (e.g. C2_Garry) has changed since sync_min_timestamp

Now obviously, with a lot of users this does not scale well, because there would be too many syncfilter rows in the database.
To overcome this issue, we introduced a graph-table to keep track of groups that users belong to. (and yes, the changes in the graph are also tracked!)

So you can assign Garry and Andy to a group "Usergroup01" and create a syncfilter for the group.
That already dramatically reduces the number of syncfilter rows.

Additionally, there may be different reasons for why a user should receive a specific record.
E.g. Garry should receive C1 because he is the creator of the C1 record
Additionally, Garry is set as "key account manager" of C1

So there are basically 2 reasons why Garry should receive C1.
When one of those reasons goes away (e.g. Andy is the new "key account manager" of C1, the system needs to know whether the syncrule can still be kept in place or if it has to be removed.
Therefore, for each sync reason - a separate syncrow would be required. So in this sample, this would normally result in 2 syncrows!
C1_Garry_OwnerOfRecord
C1_Garry_IsKeyAccountManager

Again, we need to keep the number of syncrows as small as possible.
Therefore we introduced a bigint column "AppliedRules", which is used as a Flag-Enum store.

[Flags]
public enum CustomerSyncReason `
{
   OwnerOfRecord = 1,
   IsKeyAccountManager = 2,
   SomeOtherReason = 4
}

so that way you can store up to 64 reasons per syncrow for why it should still exist.

If "AppliedRules" is zero (0), then the record C1 will be sent as toombstoned to the client.

As you see, this ruleset can be applied to almost any domain.
However, those rules have to be calculated by the application using its domain logic.
I don't see how this can be achieved by dotmim.sync!

**Comment by captmomo** [comment]
Does adding a filter require deprovisioning the server?
